Allow Failing Transactions

viquezclaudio opened the pull request:
This is an implementation of #905 
- Allow failing transactions
- Transactions are now wrapped in an ExecutedTransaction structure that conveys the result of applying such transaction

viquezclaudio committed:
Failing txn implementation

In this commit 'Executed Transaction' are instroduced.
They represent the execution result of applying a given transaction into a blockchain.
In other words, Transactions can now either succeed or fail.
When a transaction fails, the fee is deducted from the sender account

sisou commented on rpc-interface/src/types.rs:
```diff
@@ -443,6 +461,33 @@ impl From<nimiq_block::ForkProof> for ForkProof {
     }
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum ExecutedTransaction {
+    Ok(Transaction),
+    Err(Transaction),
+}
```
This is a major API change again. With this, every transaction will be wrapped into a `Ok: { ...transaction }` object (or `Err: {...}`.

I propose to serde-flatten this enum, so it doesn't wrap. But at the same time, the execution status needs to be conveyed also through the API, so this needs to be "keyed" again, so that the execution status becomes a property on the transaction object, e.g. `executionResult: bool`.

sisou commented on rpc-interface/src/types.rs:
```diff
@@ -281,12 +281,32 @@ impl Block {
-                                        )
+                                        // We obtain an internal executed transaction
+                                        // We need to grab the internal transaction and map it to the RPC transaction structure
+                                        match tx {
```
I think you can replace this with the `from_blockchain` method of the `ExecutedTransaction` type? There the two states are already differentiated.

sisou commented on rpc-interface/src/types.rs:
```diff
@@ -443,6 +463,36 @@ impl From<nimiq_block::ForkProof> for ForkProof {
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ExecutedTransaction {
+    transaction: Transaction,
```
Still needs a serde flatten here for the transaction property, to not be nested.

nibhar commented on block-production/tests/mod.rs:
```diff
@@ -431,6 +436,564 @@ fn it_can_revert_create_staker_transaction() {
+            .balance(),
+        Coin::from_u64_unchecked(1000000000000)
+    );
+
```
It would probably make sense to check that the state_root and history_root are the ones we would expect them to be. That would cover the account having the correct funds, but also that the staking contract is in the same state and so on.
Probably applies to other tests, too.

nibhar commented on primitives/account/src/basic_account.rs:
```diff
@@ -209,6 +210,105 @@ impl AccountTransactionInteraction for BasicAccount {
+            fee: transaction.fee,
+        }];
+        // If the new balance is zero, it means this account didnt exist before, so we don't need to create it.
+        if new_balance == Coin::ZERO {
```
Same as above:  With what kind of transaction would execution reach this code?
If there are cases we might also just check in the beginning for 0 fee and skip a couple of lines including db get/put as nothing will be mutated.

Additionally:
Do logs with 0 fee make sense? Only commented on this once, but there are more occurrences.

viquezclaudio replied:
Same as above, signaling, but I guess this can be safely removed.
I think for consistency we should allow logs with 0 fee.

nibhar commented on primitives/account/src/logs.rs:
```diff
@@ -333,6 +385,14 @@ pub struct BatchInfo {
 }
 
 impl BatchInfo {
+    pub fn new_empty() -> Self {
```
`Default`?

viquezclaudio replied:
just realized this is dead code, is not used

nibhar commented on blockchain/src/history/extended_transaction.rs:
```diff
@@ -128,10 +129,17 @@ impl ExtendedTransaction {
+            ExtTxData::Basic(tx) => match tx {
+                ExecutedTransaction::Ok(txn) => txn.hash(),
+                ExecutedTransaction::Err(txn) => txn.hash(),
+            },
```
Probably makes sense to move some of these into the `ExecutedTransaction` struct.
```rust
impl ExecutedTransaction {
    pub fn hash(&self) -> Blake2bHash {..}
}
```

nibhar commented on blockchain/src/history/history_store.rs:
```diff
@@ -204,6 +204,10 @@ impl HistoryStore {
+                    let tx = match tx {
+                        nimiq_transaction::ExecutedTransaction::Ok(txn) => txn,
+                        nimiq_transaction::ExecutedTransaction::Err(txn) => txn,
+                    };
```
```suggestion
                    let tx = tx.get_raw_transaction();
```

nibhar commented on blockchain/src/history/history_store.rs:
```diff
@@ -813,6 +817,11 @@ impl HistoryStore {
+                    nimiq_transaction::ExecutedTransaction::Ok(txn) => txn,
+                    nimiq_transaction::ExecutedTransaction::Err(txn) => txn,
+                };
+
```
```suggestion
                let tx = tx.get_raw_transaction();
```

nibhar commented on mempool/src/mempool.rs:
```diff
@@ -389,6 +390,17 @@ impl Mempool {
                                 .filter(|hash| {
                                     let old_tx = mempool_state.get(hash).unwrap();
 
+                                    // The sender must be able to at least pay the fee (in case the tx fails)
+                                    // (Assuming that all pending txns in the mempool for this sender are included in a block)
+                                    if !sender_account.can_fee_be_paid(
+                                        old_tx,
+                                        new_total + old_tx.fee,
+                                        blockchain.timestamp(),
+                                    ) {
+                                        // If the fee cannot be paid for this transaction, we filter it
+                                        return true;
+                                    }
+
```
Is this even necessary? The following lines check if `old_tx.total_value()` is still covered, which by construction is always >= `old_tx.fee`

viquezclaudio replied:
That is true, the _total_value_ check is stronger than the fee check because _total_value_ already incudes the fee, which seems to be sufficient for basic transactions.
But is this sufficient for other transaction types? i.e: vesting and htlc contracts?

nibhar commented on mempool/src/verify.rs:
```diff
@@ -274,6 +276,19 @@ pub(crate) async fn verify_tx<'a>(
         recipient_current_balance -= recipient_state.total;
     }
 
+    // The sender must be able to at least pay the fee (in case the tx fails), assumming all pending txns in the mempool for this sender are included in a block
+    if !sender_account.can_fee_be_paid(
+        transaction,
+        sender_current_balance + transaction.fee,
+        blockchain.timestamp(),
+    ) {
+        // If the fee cannot be paid for this transaction, we reject it.
+        return Err(VerifyErr::CantPayFee);
+    }
```
Same here.

viquezclaudio replied:
This check is designed for vesting/htlc contracts, because apart from just verifying if the fee can be paid from a balance perspective, in the case of vesting/htlc we need to make sure the fee can actually be deducted from the contract (contract already vested with enough funds, etc)

nibhar commented on primitives/account/src/account.rs:
```diff
@@ -73,6 +73,16 @@ impl Account {
+        block_time: u64,
+    ) -> bool {
+        self.can_pay_fee(transaction, mempool_balance, block_time)
+    }
```
can we not just use `AccountTransactionInteraction::can_pay_fee`?

nibhar commented on primitives/account/src/accounts.rs:
```diff
@@ -202,19 +470,64 @@ impl Accounts {
-            timestamp,
-            recipient_receipts,
-        )?;
+        for (index, transaction) in transactions.iter().enumerate() {
```
Does this not have to be reverse order, too?

nibhar commented on primitives/account/src/basic_account.rs:
```diff
@@ -209,6 +210,105 @@ impl AccountTransactionInteraction for BasicAccount {
+                if transaction.total_value() != Coin::ZERO {
+                    None
+                } else {
+                    Some(Account::Basic(BasicAccount {
+                        balance: Coin::ZERO,
+                    }))
```
More out of interest: What would be a transaction leading to this code being executed?

viquezclaudio replied:
This code is inherited from _commit_outgoing_transaction_ and I think it is used for signaling transactions: i.e: if the transaction total value is 0 and the account does not exist, we just use a "mock" basic account.
It was added in commit _a5ed6b79d7ca2883fe19b28e8ca8da6aaafdb294_ and if I remember correctly it was added because there was some panic or some issue that we were experimenting, however I don't know exactly what a signaling transaction that fails means, so if that is never a possibility, we can just remove this

nibhar commented on primitives/account/src/staking_contract/traits.rs:
```diff
@@ -436,6 +436,96 @@ impl AccountTransactionInteraction for StakingContract {
+        _current_balance: Coin,
+        _block_time: u64,
+    ) -> bool {
+        // TODO Implementation missing: we need to support the ability to pay the fee from a third party for outgoing stacking transactions
+        // When that is ready, we can use this function to verify if the third party can actually pay the fee
+        // For now we just return true, to be consistant with the current implementation
```
For `Unstake` we should be able to calculate this?

viquezclaudio replied:
yes good catch!
The only problem is that I need the _accounts_tree_  to get the staking contract in order to obtain the staker balance, but I do not have the accounts_tree in this interface, is there an alternative way to do this?

nibhar commented on primitives/account/tests/accounts.rs:
```diff
@@ -801,3 +815,124 @@ fn accounts_performance_history_sync_batches_many_to_many() {
+
+    db_txn.commit();
+
+    // We should deduct the fee from the staking contract balance
```
```suggestion
    // We should deduct the fee from the vesting contract balance
```

nibhar commented on primitives/account/src/vesting_contract.rs:
```diff
@@ -256,6 +256,130 @@ impl AccountTransactionInteraction for VestingContract {
+        };
+
+        // Note that in this type of transactions the fee is paid (deducted) from the contract balance
+        let new_balance = Account::balance_sub(account.balance(), transaction.fee)?;
```
Do we not have to check `min_cap` here as well as the correctness of the recipients signature?
Both of these checks are performed for `commit_outgoing_transaction`.

viquezclaudio replied:
This is covered by _can_pay_fee_ , the mempool before accepting the transaction needs to call _can_pay_fee_ which would check that the transaction signer is contract owner.
Do you need think it is necessary to also add this check here?

nibhar commented on primitives/block/src/block.rs:
```diff
@@ -231,7 +231,15 @@ impl Block {
+                .map(|ex| {
+                    ex.transactions
+                        .iter()
+                        .map(|tx| match tx {
+                            ExecutedTransaction::Ok(tx) => tx.fee,
+                            ExecutedTransaction::Err(tx) => tx.fee,
+                        })
```
I'd say either
```suggestion
                        .map(|tx| tx.get_raw_transaction().fee()
```
or add `ExecutedTransaction::fee(&self)`

nibhar commented on primitives/block/src/micro_block.rs:
```diff
@@ -75,7 +76,20 @@ pub struct MicroBody {
+        // Extract the transactions from the block
+        self.transactions
+            .iter()
+            .map(|txn| match txn {
+                ExecutedTransaction::Ok(txn) => txn.clone(),
+                ExecutedTransaction::Err(txn) => txn.clone(),
+            })
```
```suggestion
            .map(|txn| txn.get_raw_transaction().clone())
```

viquezclaudio committed:
Fix receipt ordering

During a revert operation, receipts are reverse ordered.
Transactions were iterated in the order they were applied, which is problematic
since the receipts would not match their corresponding transactions.
With this change, both receipts and transactions are reverse ordered before reverting them,
so both receipts and transactions match

viquezclaudio committed:
Various bug fixes and enhancements

- Removed transaction ordering in blocks
- Fixed reverting contract creation txns
- remove can_fee_be_paid wrapper
- renamed get_root_with
- some test fixes

viquezclaudio committed:
Refactor delete validator transaction

- Make several changes to the delete validator transaction to make it compatible
  with failing transactions
- The validator deposit can be reduced by sending failing outgoing staking transactions

Saiv46 commented:
Do I miss something? How and why do we need to handle failed transactions? If account tries to move more funds that it has, transaction will be rejected and new transaction will be needed.

viquezclaudio commented:
Correct, we perform balance checks in the mempool.
This is more intended for staking contract misusage (trying to delete a validator that no longer exists, etc)
